### ISSUE TITLE
Blocked by constant issues with node-gyp-build

### ISSUE THREAD

**Issue by rafaelfaria** [aws-amplify/amplify-cli#11292]
### Before opening, please confirm:

- [X] I have installed the latest version of the Amplify CLI (see above), and confirmed that the issue still persists.
- [X] I have [searched for duplicate or closed issues](https://github.com/aws-amplify/amplify-cli/issues?q=is%3Aissue+).
- [X] I have read the guide for [submitting bug reports](https://github.com/aws-amplify/amplify-cli/blob/dev/CONTRIBUTING.md#bug-reports).
- [X] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.
- [X] I have removed any sensitive information from my code snippets and submission.

### How did you install the Amplify CLI?

npm

### If applicable, what version of Node.js are you using?

v16.17.0

### Amplify CLI Version

10.3.2

### What operating system are you using?

Mac

### Did you make any manual changes to the cloud resources managed by Amplify? Please describe the changes made.

No manual changes made

### Amplify Categories

Not applicable

### Amplify Commands

push

### Describe the bug

Ever since I updated to the latest version of amplify I have been getting into issues with either symlinks or node-gyp-build.

Lately, I have been using a lambda function that uses web3, however, I have come to so many issues, that I decided to move away from it. Even locally, where[ this suggestion ](https://github.com/aws-amplify/amplify-cli/issues/10471#issuecomment-1176559231)should've worked, it's not working anymore.

I thought moving away from web3 would have worked. However, I still keep getting into issues when deploying my lambda layer. Here it's the latest message:

```
2022-10-30T12:49:40.241Z [INFO]: 🛑 Packaging lambda function failed with the error
                                 Command failed with exit code 127: yarn --no-bin-links --production
                                 warning package.json: No license field
                                 warning No license field
                                 error /codebuild/output/src759146275/src/myapp/amplify/backend/function/cryptoappCommonLayer/lib/nodejs/node_modules/bufferutil: Command failed.
                                 Exit code: 127
                                 Command: node-gyp-build
                                 Arguments:
                                 Directory: /codebuild/output/src759146275/src/myapp/amplify/backend/function/cryptoappCommonLayer/lib/nodejs/node_modules/bufferutil
                                 Output:
                                 /bin/sh: node-gyp-build: command not found
                                 yarn install v1.22.0
                                 [1/4] Resolving packages...
                                 [2/4] Fetching packages...
                                 [3/4] Linking dependencies...
                                 [4/4] Building fresh packages...
                                 info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
2022-10-30T12:49:40.241Z [INFO]: Learn more at: https://docs.amplify.aws/cli/project/troubleshooting/
```

I don't have bufferutils on my package so it might be some sort of dependency from one of the packages.

The issue is SO bizarre, that the SAME code, everything EXACTLY the same, merged into 2 branches, but running into 2 different environments (one for prod and another one for dev). Gives a completely different issue

```
2022-10-31T12:00:37.138Z [INFO]: [0mSuggested configuration for new layer versions:[0m
2022-10-31T12:00:37.138Z [INFO]: 
2022-10-31T12:00:37.140Z [INFO]: [0mcryptoappCommonLayer[0m
                                 [0m  - Access permissions: Maintain existing permissions[0m
                                 [0m  - Description: Updated layer version  2022-10-31T12:00:37.138Z[0m
2022-10-31T12:00:37.140Z [INFO]: 
2022-10-31T12:00:56.049Z [WARNING]: ✖ There was an error initializing your environment.
2022-10-31T12:00:56.173Z [INFO]: 🛑 Packaging lambda function failed with the error
                                 Command failed with exit code 127: yarn --no-bin-links --production
                                 warning package.json: No license field
                                 warning No license field
                                 error /codebuild/output/src773265498/src/myapp/amplify/backend/function/cryptoappCommonLayer/lib/nodejs/node_modules/utf-8-validate: Command failed.
                                 Exit code: 127
                                 Command: node-gyp-build
                                 Arguments:
                                 Directory: /codebuild/output/src773265498/src/myapp/amplify/backend/function/cryptoappCommonLayer/lib/nodejs/node_modules/utf-8-validate
                                 Output:
                                 /bin/sh: node-gyp-build: command not found
                                 yarn install v1.22.0
                                 [1/4] Resolving packages...
                                 [2/4] Fetching packages...
                                 [3/4] Linking dependencies...
                                 [4/4] Building fresh packages...
                                 info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
2022-10-31T12:00:56.173Z [INFO]: Learn more at: https://docs.amplify.aws/cli/project/troubleshooting/
```

### Expected behavior

I would expect to be deployed without me worrying if there is a symlink or not. Or the requirement of.a node-gyp-build.

### Reproduction steps

I actually not sure how to reproduce this because I am not sure how to narrow down the packaged which needs to be installed in order to get this error.

### GraphQL schema(s)

<details>

```graphql
# Put schemas below this line


```

</details>

### Log output

<details>

```
# Put your logs below this line


```

</details>

**Comment by rafaelfaria** [comment]
Hi @edwardfoyle  please a bit of help with this. We are completely blocked by it. 

We used to have a lambda function calling another lambda function which never caused any problems, however, it is VERY slow. By moving everything to our lambda layer, we made 6x faster all the calls and responses.

A bit of further investigation, we narrow down the package [telegram](https://github.com/gram-js/gramjs) being the problem. After I removed it from the lambda layer it built it. However, if the same package is on a lambda function instead, I get no problems whatsoever.

This is really blocking my release. Help please?

**Comment by josefaidt** [comment]
Hey @rafaelfaria :wave: thanks for raising this! What version of Node is being used in the Amplify Hosting console? By default it is set to 14, and I'm able to reproduce this error by downgrading to 14 locally, but not on 16 or 18 as you have mentioned. Can you try setting the Node version in Amplify Hosting to 16 or higher?

**Comment by rafaelfaria** [comment]
@josefaidt  You are right that I am using node 14 in the console.

```
ENV VERSION_NODE_DEFAULT=$VERSION_NODE_14
```

How do I change that? I tried to do it through the build panel, redeploy but doesn't seem to work, it still set the VERSION_NODE_14 as default!
[image](https://user-images.githubusercontent.com/227917/199367772-dbaeb73c-e9e1-4270-be35-5d3665a71357.png)
![image](https://user-images.githubusercontent.com/227917/199367898-78720b4a-a7ab-4602-ba7c-03ed00d08c3f.png)

**Comment by rafaelfaria** [comment]
More updates on this.

After setting my node to 16, I actually noticed the version I chose gets installed during the `Build -> Cloning Repository` step. 🤷‍♂️ 

However I am getting the following error:

```
2022-11-02T10:35:38.463Z [INFO]: # Node version 16 is available for installation
2022-11-02T10:35:38.559Z [INFO]: # Installing Node version 16
2022-11-02T10:35:46.273Z [INFO]: # Now using Node version 16
2022-11-02T10:35:47.684Z [WARNING]: Cannot find module '/root/.nvm/versions/node/v16.18.0/lib/node_modules/@aws-amplify/cli/package.json'
                                    Require stack:
                                    - /snapshot/AWSMobilePioneerExecute/out/index.js

```

AND, still throws a similar error, although, besides the problem listed in the description of my issue, it also throws some warnings to even more stuff now (it does not do it on my local) but I think its due the node version and deprecation, it shouldn't fail now.

```
2022-11-02T10:37:29.320Z [INFO]: [0mSuggested configuration for new layer versions:[0m
2022-11-02T10:37:29.322Z [INFO]: [0mcryptoappCommonLayer[0m
                                 [0m  - Access permissions: Maintain existing permissions[0m
                                 [0m  - Description: Updated layer version  2022-11-02T10:37:29.319Z[0m
2022-11-02T10:37:29.322Z [INFO]: 
2022-11-02T10:37:49.575Z [WARNING]: ✖ There was an error initializing your environment.
2022-11-02T10:37:49.667Z [INFO]: 🛑 Packaging lambda function failed with the error
                                 Command failed with exit code 127: yarn --no-bin-links --production
                                 warning package.json: No license field
                                 warning No license field
                                 warning node-telegram-bot-api > request@2.88.2: request has been deprecated, see https://github.com/request/request/issues/3142
                                 warning node-telegram-bot-api > request-promise@4.2.6: request-promise has been deprecated because it extends the now deprecated request package, see https://github.com/request/request/issues/3142
                                 warning node-telegram-bot-api > request > har-validator@5.1.5: this library is no longer supported
                                 warning node-telegram-bot-api > request > uuid@3.4.0: Please upgrade  to version 7 or higher.  Older versions may use Math.random() in certain circumstances, which is known to be problematic.  See https://v8.dev/blog/math-random for details.
                                 error /codebuild/output/src203981451/src/myapp/amplify/backend/function/cryptoappCommonLayer/lib/nodejs/node_modules/bufferutil: Command failed.
                                 Exit code: 127
                                 Command: node-gyp-build
                                 Arguments:
                                 Directory: /codebuild/output/src203981451/src/myapp/amplify/backend/function/cryptoappCommonLayer/lib/nodejs/node_modules/bufferutil
                                 Output:
                                 /bin/sh: node-gyp-build: command not found
                                 yarn install v1.22.0
                                 info No lockfile found.
                                 [1/4] Resolving packages...
                                 [2/4] Fetching packages...
                                 [3/4] Linking dependencies...
                                 [4/4] Building fresh packages...
                                 info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
2022-11-02T10:37:49.668Z [INFO]: Learn more at: https://docs.amplify.aws/cli/project/troubleshooting/
2022-11-02T10:37:49.868Z [WARNING]: - Uploading files.
```

**Comment by rafaelfaria** [comment]
So... More on this @josefaidt. 

As you mentioned on your response, I was indeed on node 16.17.0 on my local, and I was able to push it without any issues. However, I just realised that by following the thread https://github.com/aws-amplify/amplify-cli/issues/10471 I end up installing `node-gyp-build` globally. However, I have since uninstalled it, and now, even on the latest node 16, I am experiencing the issue mentioned above.

I then, reinstalled `sudo npm i -g node-gyp-build` on my local and now I managed to make it work again. That makes me wonder if the node version is indeed the root cause of this not working on the layers. And also, the node version doesn't explain why it would work on lambda functions but not layers.

As I mentioned in my description of the issue, I had this working fine on lambda functions and only started getting errors when I moved it to the layer.

Any other ideas?

**Comment by josefaidt** [comment]
Hey @rafaelfaria 

```
2022-11-02T10:35:38.463Z [INFO]: # Node version 16 is available for installation
2022-11-02T10:35:38.559Z [INFO]: # Installing Node version 16
2022-11-02T10:35:46.273Z [INFO]: # Now using Node version 16
2022-11-02T10:35:47.684Z [WARNING]: Cannot find module '/root/.nvm/versions/node/v16.18.0/lib/node_modules/@aws-amplify/cli/package.json'
                                    Require stack:
                                    - /snapshot/AWSMobilePioneerExecute/out/index.js
```

this appears we will want to reinstall the CLI after bumping the node version

```
nvm install 16
npm i -g @aws-amplify/cli
```

I don't believe installing `node-gyp-build` should be a requirement (P.S. it would also not be recommended to install npm packages with sudo). Is there a prebuild/build script in your function that is relying on this dependency? 

It looks like this is a transient dependency and should appear in the lockfile under `bufferutil` as shown in your stack trace

```
bufferutil@^4.0.1, bufferutil@^4.0.3:
  version "4.0.7"
  resolved "https://registry.npmjs.org/bufferutil/-/bufferutil-4.0.7.tgz#60c0d19ba2c992dd8273d3f73772ffc894c153ad"
  integrity sha512-kukuqc39WOHtdxtw4UScxF/WVnMFVSQVKhtx3AjZJzhd0RGZZldcrfSEbVsWWe6KNH253574cq5F+wpv0G9pJw==
  dependencies:
    node-gyp-build "^4.3.0"
```

can you try deleting the layers node_modules and re-attempt the push on Node 16 without `node-gyp-build` installed globally?

**Comment by josefaidt** [comment]
After I posted this I switched node versions to 18 and started receiving this error on push

```
🛑 Packaging lambda function failed with the error 
Command failed with exit code 127: yarn --no-bin-links --production
warning package.json: No license field
warning No license field
error /Users/josef/Documents/projects/aws-amplify/reproductions/11278/amplify/backend/function/11278layer4ec2e6f1/lib/nodejs/node_modules/utf-8-validate: Command failed.
Exit code: 127
Command: node-gyp-build
Arguments: 
Directory: /Users/josef/Documents/projects/aws-amplify/reproductions/11278/amplify/backend/function/11278layer4ec2e6f1/lib/nodejs/node_modules/utf-8-validate
Output:
/bin/sh: node-gyp-build: command not found
yarn install v1.22.19
[1/4] Resolving packages...
[2/4] Fetching packages...
[3/4] Linking dependencies...
[4/4] Building fresh packages...
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command
```

**Comment by rafaelfaria** [comment]
Hi @josefaidt 
first of all, thanks for all the assistance. I really appreciate it.

> After I posted this I switched node versions to 18 and started receiving this error on push
> 
> ```
> 🛑 Packaging lambda function failed with the error 
> Command failed with exit code 127: yarn --no-bin-links --production
> warning package.json: No license field
> warning No license field
> error /Users/josef/Documents/projects/aws-amplify/reproductions/11278/amplify/backend/function/11278layer4ec2e6f1/lib/nodejs/node_modules/utf-8-validate: Command failed.
> Exit code: 127
> Command: node-gyp-build
> Arguments: 
> Directory: /Users/josef/Documents/projects/aws-amplify/reproductions/11278/amplify/backend/function/11278layer4ec2e6f1/lib/nodejs/node_modules/utf-8-validate
> Output:
> /bin/sh: node-gyp-build: command not found
> yarn install v1.22.19
> [1/4] Resolving packages...
> [2/4] Fetching packages...
> [3/4] Linking dependencies...
> [4/4] Building fresh packages...
> info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command
> ```

Does it mean it would be considered a bug since it works on a lambda function but not on a layer? or is this design to work like that?

I also agree that we shouldn't be installing npm packages with sudo, however, without that, I am completely blocked with my project. This is the only way I can actually deploy at the moment since CD is failing because of the issue.

> can you try deleting the layers node_modules and re-attempt the push on Node 16 without node-gyp-build installed globally?

To report back, I have deleted the node_modules, yarn.lock, and removed the global node-gyp-build and I am still getting the issue. It doesn't sound like the node version is a factor here.

Any other ideas you would like me to try?

**Comment by josefaidt** [comment]
Hey @rafaelfaria I'm able to reproduce this consistently without a lockfile present in the Lambda layer, however after I `cd` into that directory and run `yarn` to install deps I'm able to run `amplify push -y` without issue (on Node 18). From there I can also switch Node versions down to 16, create another file in the layer (to trigger an update) and push.

Can you try regenerating the lockfile in the affected layer?

**Comment by rafaelfaria** [comment]
Hi @josefaidt 
I have just tested the scenario you suggested and indeed worked. Having the yarn.lock in the folder does allow me to push things locally without any errors. 🎉 

However, my main issue isn't so much the local scenario, which I had the workaround to install the node-gyp-build globally, but rather the CI/CD in the cloud. Upon following what you suggest, and pushing things to my branch, amplify console is still complaining about the issue. How do we solve that? Did you manage to get a successful build?

>this appears we will want to reinstall the CLI after bumping the node version
>```
>nvm install 16
>npm i -g @aws-amplify/cli
>```
How do we make this happen since I don't have the control of installing the aws-amplify/cli aftwards? Do you think this solves the problem?

**Comment by montycheese** [comment]
Having this issue too when it comes to CI/ID. Any update solution here?

**Comment by josefaidt** [comment]
> However, my main issue isn't so much the local scenario, which I had the workaround to install the node-gyp-build globally, but rather the CI/CD in the cloud. Upon following what you suggest, and pushing things to my branch, amplify console is still complaining about the issue. How do we solve that? Did you manage to get a successful build?

To clarify, are you still experiencing this issue in the Amplify Hosting builds after creating a lockfile for the affected layer/function? 

> How do we make this happen since I don't have the control of installing the aws-amplify/cli aftwards? Do you think this solves the problem?

In your buildspec (`amplify.yml`) file, add these lines to the `preBuild` phase before `amplifyPush --simple`. I would recommend ensuring the Node and Amplify CLI versions used by Amplify Hosting matches what you currently have installed locally.

**Comment by rafaelfaria** [comment]
Hi @josefaidt 

> To clarify, are you still experiencing this issue in the Amplify Hosting builds after creating a lockfile for the affected layer/function?

Locally this indeed solved the problem, however, Amplify hosting still breaking.


> In your buildspec (`amplify.yml`) file, add these lines to the `preBuild` phase before `amplifyPush --simple`. I would recommend ensuring the Node and Amplify CLI versions used by Amplify Hosting matches what you currently have installed locally.

I tried modifying the `amplify.yml`, and i didn't see any changes, then I tried to edit it through the build settings, and the problem still persists.

Here is my `amplify.yml`. However, i don't think its triggering the prebuild

```
version: 1
backend:
  phases:
    preBuild:
      commands:
        - nvm install 16
        - npm i -g @aws-amplify/cli
    build:
      commands:
        - '# Execute Amplify CLI with the helper script'
        - amplifyPush --simple
frontend:
  phases:
    preBuild:
      commands:
        - yarn install
    build:
      commands:
        - yarn run build
  artifacts:
    baseDirectory: build
    files:
      - '**/*'
  cache:
    paths:
      - node_modules/**/*

```

When the build happens I still get this in the "Cloning repository" phase:

``` 
2022-11-08T06:26:14.843Z [INFO]: # Node version 16 is available for installation
2022-11-08T06:26:15.011Z [INFO]: # Installing Node version 16
2022-11-08T06:26:22.099Z [INFO]: # Now using Node version 16
2022-11-08T06:26:23.183Z [WARNING]: Cannot find module '/root/.nvm/versions/node/v16.18.1/lib/node_modules/@aws-amplify/cli/package.json'
                                    Require stack:
                                    - /snapshot/AWSMobilePioneerExecute/out/index.js
2022-11-08T06:26:23.183Z [INFO]: # Patching NPM package '@aws-amplify/cli' to 10.4.0...
2022-11-08T06:26:30.589Z [INFO]: # Done patching NPM package '@aws-amplify/cli' to version 10.4.0
```

Then I still get the ERROR.

Did you manage to get it working? I'm a bit clueless on what to do next.

**Comment by rafaelfaria** [comment]
@linglongguzi the suggestion @josefaidt  mentioned above works for me. You need to yarn in the src folder (I'm using node 16).

However, still doesn't work in the console. I'm waiting @josefaidt  to give me some guidance here.

**Comment by geh18** [comment]
Same issue here with `node-pre-gyp`. The package is built successfully when pushed with `amplify push` but when the Amplify CI/CD triggers then the follow error code 127 is issued.

**Comment by josefaidt** [comment]
Hey @rafaelfaria apologies for the delay here!! I'm also able to reproduce this issue with an updated lockfile but missing node_modules, which leads me to believe these dependencies are relying on a binary available during dependency install, but only if we have installed devdependencies already. As a workaround can you try to run `yarn` in the functions `src` directory prior to `amplifyPush --simple`?

**Comment by rafaelfaria** [comment]
@josefaidt  no worries. Thanks for getting back to me.

do you have an example of how I cd into the function folder? what is the path I need to cd to from the script?
ex: /opt/amplify/functions/?

Another question is, do i need to return to the root to run the amplifyPush? 

appreciate the help navigating through this update.

**Comment by josefaidt** [comment]
Hey @rafaelfaria sorry about that, I'm thinking something along the lines of:

```
cd amplify/backend/function/<function-name>/src && yarn && cd -
```

> Another question is, do i need to return to the root to run the amplifyPush?

While the CLI has functionality to look up the directory tree to determine the root of your Amplify project I would still suggest moving back to the project root, and `cd -` should take care of this for us

**Comment by rafaelfaria** [comment]
Hi @josefaidt 
Thanks for helping me out with this.

I did what you suggested. Added one step at a time so I could see what was going on.

```
    # Starting phase: build
                                 # Executing command: cd amplify/backend/function/myAppLayer/lib/nodejs
2022-11-27T04:27:12.402Z [INFO]: # Executing command: ls -la
2022-11-27T04:27:12.404Z [INFO]: total 76
                                 drwxr-xr-x 2 root root    61 Nov 27 04:26 .
                                 drwxr-xr-x 3 root root    20 Nov 27 04:26 ..
                                 -rw-r--r-- 1 root root   573 Nov 27 04:26 package.json
                                 -rw-r--r-- 1 root root    39 Nov 27 04:26 README.txt
                                 -rw-r--r-- 1 root root 67216 Nov 27 04:26 yarn.lock
2022-11-27T04:27:12.404Z [INFO]: # Executing command: yarn
2022-11-27T04:27:12.689Z [INFO]: yarn install v1.22.0
2022-11-27T04:27:12.708Z [WARNING]: warning package.json: No license field
2022-11-27T04:27:12.741Z [WARNING]: warning No license field
2022-11-27T04:27:12.744Z [INFO]: [1/4] Resolving packages...
2022-11-27T04:27:12.820Z [INFO]: [2/4] Fetching packages...
2022-11-27T04:27:17.265Z [INFO]: [3/4] Linking dependencies...
2022-11-27T04:27:19.304Z [INFO]: [4/4] Building fresh packages...
2022-11-27T04:27:19.499Z [INFO]: Done in 6.81s.
2022-11-27T04:27:19.515Z [INFO]: # Executing command: ls -la
2022-11-27T04:27:19.517Z [INFO]: total 88
                                 drwxr-xr-x   3 root root    81 Nov 27 04:27 .
                                 drwxr-xr-x   3 root root    20 Nov 27 04:26 ..
                                 drwxr-xr-x 188 root root  8192 Nov 27 04:27 node_modules
                                 -rw-r--r--   1 root root   573 Nov 27 04:26 package.json
                                 -rw-r--r--   1 root root    39 Nov 27 04:26 README.txt
                                 -rw-r--r--   1 root root 67216 Nov 27 04:26 yarn.lock
2022-11-27T04:27:19.517Z [INFO]: # Executing command: cd -
                                 /codebuild/output/src148606459/src/myApp
```

However, that doesn't seem to actually fix the issue. 

```
2022-11-27T04:28:17.424Z [INFO]: 
2022-11-27T04:28:34.603Z [WARNING]: ✖ There was an error initializing your environment.
2022-11-27T04:28:34.612Z [INFO]: 🛑 Packaging lambda function failed with the error
                                 Command failed with exit code 127: yarn --no-bin-links --production
                                 warning package.json: No license field
                                 warning No license field
                                 error /codebuild/output/src148606459/src/myApp/amplify/backend/function/myAppLayer/lib/nodejs/node_modules/bufferutil: Command failed.
                                 Exit code: 127
                                 Command: node-gyp-build
                                 Arguments:
                                 Directory: /codebuild/output/src148606459/src/myApp/amplify/backend/function/myAppLayer/lib/nodejs/node_modules/bufferutil
                                 Output:
                                 /bin/sh: node-gyp-build: command not found
                                 yarn install v1.22.0
                                 [1/4] Resolving packages...
                                 [2/4] Fetching packages...
                                 [3/4] Linking dependencies...
                                 [4/4] Building fresh packages...
                                 info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
2022-11-27T04:28:34.613Z [INFO]: Learn more at: https://docs.amplify.aws/cli/project/troubleshooting/
```

Did you actually manage to fix the issue on your end? Not quite sure what else to try.

Also, meant to ask, why would this happens to a layer but not to a function?

**Comment by rafaelfaria** [comment]
hey @josefaidt 
I noticed this was added to To do, what does it exactly mean?
Also, are there any other workaround ideas that I can try? I'm blocked from automatically CD/CI my prod environment. It needs to always do it manually.

**Comment by josefaidt** [comment]
Hey @rafaelfaria apologies for the delay and for the confusion. This was added to the "function build redesign" project board to track this issue alongside others as we aim to improve the function experience (including Layers).

> Did you actually manage to fix the issue on your end? Not quite sure what else to try.

For a single Lambda function this is reproducible by intentionally not including the lockfile, including an out-of-date lockfile (i.e. without references to this dep), or by pushing without devdependencies installed in the function's directory.

For a Lambda Layer this is reproducible with the Amplify Hosting setup described here.

By adding `telegram` as a dependency to the layer and immediately attempting to push will result in the error shown here:

```
{
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "dependencies": {
    "telegram": "^2.14.7"
  },
  "devDependencies": {}
}
```

```
🛑 Packaging lambda function failed with the error 
Command failed with exit code 127: yarn --no-bin-links --production
warning package.json: No license field
warning No license field
error /Users/josef/Documents/projects/aws-amplify/reproductions/funwithlayers/amplify/backend/function/funwithlayerslayer8ee655b2/lib/nodejs/node_modules/utf-8-validate: Command failed. 
```

Here we can see the failed execution attempt of `yarn --no-bin-links --production` when running this command locally and `node_modules` does not exist, which is what motivated the request to install dependencies manually in the Hosting pipeline

```
funwithlayerslayer8ee655b2/lib/nodejs 
➜  yarn --no-bin-links --production
yarn install v1.22.19
warning package.json: No license field
warning No license field
[1/4] 🔍  Resolving packages...
success Already up-to-date.
✨  Done in 0.05s.

funwithlayerslayer8ee655b2/lib/nodejs 
➜  rm -rf node_modules/

funwithlayerslayer8ee655b2/lib/nodejs 
➜  yarn --no-bin-links --production
yarn install v1.22.19
warning package.json: No license field
warning No license field
[1/4] 🔍  Resolving packages...
[2/4] 🚚  Fetching packages...
[3/4] 🔗  Linking dependencies...
[4/4] 🔨  Building fresh packages...
⠁ 
⠁ 
error /Users/josef/Documents/projects/aws-amplify/reproductions/funwithlayers/amplify/backend/function/funwithlayerslayer8ee655b2/lib/nodejs/node_modules/utf-8-validate: Command failed.
Exit code: 127
Command: node-gyp-build
Arguments: 
Directory: /Users/josef/Documents/projects/aws-amplify/reproductions/funwithlayers/amplify/backend/function/funwithlayerslayer8ee655b2/lib/nodejs/node_modules/utf-8-validate
Output:



funwithlayerslayer8ee655b2/lib/nodejs 
➜  yarn
yarn install v1.22.19
warning package.json: No license field
warning No license field
[1/4] 🔍  Resolving packages...
[2/4] 🚚  Fetching packages...
[3/4] 🔗  Linking dependencies...
[4/4] 🔨  Building fresh packages...
✨  Done in 0.75s.

funwithlayerslayer8ee655b2/lib/nodejs 
➜  yarn --no-bin-links --production
yarn install v1.22.19
warning package.json: No license field
warning No license field
[1/4] 🔍  Resolving packages...
[2/4] 🚚  Fetching packages...
[3/4] 🔗  Linking dependencies...
[4/4] 🔨  Building fresh packages...
✨  Done in 0.14s.
```

Upon further inspection it was noticed that the node_modules we install by manually editing our buildspec to include the following are removed during `amplifyPush --simple`:

```yml
version: 1
backend:
  phases:
    # IMPORTANT - Please verify your build commands
    build:
      commands:
        - '# Execute Amplify CLI with the helper script'
        - cd amplify/backend/function/funwithlayerslayer8ee655b2/lib/nodejs; yarn; cd -
        - amplifyPush --simple
frontend:
  phases:
    build:
      commands: []
  artifacts:
    # IMPORTANT - Please verify your build output directory
    baseDirectory: /
    files:
      - '**/*'
  cache:
    paths: []
```

`amplifyPush --simple` runs a headless pull command in the background which is removing the previously-installed node_modules:

```
amplify pull --appId durv2futilomz --envName dev
```

Marking as a bug 🙂 

Unfortunately the workaround is to write and use your own headless pull/init script, similar to  the script in this [gist](https://gist.github.com/josefaidt/a8cd6071d250693fdae73c84bc5b23f7) and on our [docs](https://docs.amplify.aws/cli/usage/headless/#sample-script). Please keep in mind the gist includes the smallest possible example, while the script will need to grow to support resources such as imported auth.

NOTE: this issue can be reproduced locally with the following steps:

1. `amplify init -y`
2. `amplify add function` > add layer
3. add `telegram` to the layer's `package.json` as a `dependency`
4. install dependencies for the first time by `cd`-ing into the function's `lib/nodejs` directory and running `yarn install`. This will generate the lockfile
5. run `amplify pull --appId <your-app-id> --envName <current-env-name>`
6. attempt to run `amplify push -y`
7. observe error

**Comment by jansila** [comment]
Hi, 
I'm not sure if it is off-topic, but I had the same error (works locally, cannot build in CI) when trying to use `canvas` in lambda and solved it by using a layer published here (runs Node 14x)
`https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:990551184979:applications~lambda-layer-canvas-nodejs`.

I hope it maybe helps.
Good luck!

**Comment by rafaelfaria** [comment]
Hey @josefaidt 

I have been trying to play around with the amplify-init. but I'm a bit clueless, even with the simplest example. So still blocked here :/

> while the script will need to grow to support resources such as imported auth.
What do you mean by that?

Is there any ETA on when this bug will be addressed? 

Thanks.

**Comment by rafaelfaria** [comment]
@josefaidt do you have any ETA on when this will be fixed? This is blocking our continuous integration/deployment.

I have no idea how to create the script to do the deployment. Would really appreciate the help.

**Comment by josefaidt** [comment]
Hey @rafaelfaria apologies for the delay here, and unfortunately I do not have an estimation on timelines for this fix. I have an example of a headless script, however it can become more complex depending on the project setup: https://gist.github.com/josefaidt/a8cd6071d250693fdae73c84bc5b23f7

Do you have any imported resources or OAuth credentials? If so you will need to pass those into the push script. But in short, you will want to _initialize_ your project, then _push_. More information on CLI's headless mode can be found here https://docs.amplify.aws/cli/usage/headless/

**Comment by rafaelfaria** [comment]
Hey @josefaidt 
When you say imported resources or Oauth credentials, what exactly do you mean?

My project only uses amplify resources.

Auth, Storage,Apis, Functions and Layers.

Would your script would be enough?

**Comment by josefaidt** [comment]
Hey @rafaelfaria I'm referring to resources imported to the project with `amplify import <category>`, such as [auth](https://docs.amplify.aws/cli/auth/import/) and [storage](https://docs.amplify.aws/cli/storage/import/). If the resources in your project are all managed by Amplify this script should be sufficient

**Comment by rafaelfaria** [comment]
Hey @josefaidt 
Just an update. I have tried to add the following:


amplify.yml
```yml
version: 1
backend:
  phases:
    # IMPORTANT - Please verify your build commands
    build:
      commands:
        - '# Execute Amplify CLI with the helper script'
        - cd amplify/backend/function/myappLayer/lib/nodejs/lib/nodejs; yarn; cd -
        - amplifyPush --simple
frontend:
  phases:
    build:
      commands: []
  artifacts:
    # IMPORTANT - Please verify your build output directory
    baseDirectory: /
    files:
      - '**/*'
  cache:
    paths: []

```

and the amplify-init.sh

```bash
#!/bin/bash
set -e
IFS='|'

AMPLIFY_ENVIRONMENT='dev'
AWS_REGION='us-east-1'

AWSCLOUDFORMATIONCONFIG="{\
\"configLevel\":\"project\",\
\"useProfile\":true,\
\"profileName\":\"default\",\
\"region\":\"$AWS_REGION\"\
}"
AMPLIFY="{\
\"envName\":\"$AMPLIFY_ENVIRONMENT\",\
}"

PROVIDERS="{\
\"awscloudformation\":$AWSCLOUDFORMATIONCONFIG}"

amplify init \
  --amplify $AMPLIFY \
  --providers $PROVIDERS \
  --yes || true
```

without much success. I still get the same error `/bin/sh: node-gyp-build: command not found` :/

**Comment by NoHop3** [comment]
I read the whole thread and I want to say that this is currently an issue I am facing. The package I have that is causing the issue is bcrypt. Also locally I was able to push successfully but the CI/CD from Github is failing 

```
2023-10-04T15:52:50.869Z [INFO]: # Cloning repository: git@github.com:Automating-Hydroponics-BPR2/via-sem7-bpr2-client.git
2023-10-04T15:52:52.231Z [INFO]: 
2023-10-04T15:52:52.231Z [INFO]: Cloning into 'via-sem7-bpr2-client'...
2023-10-04T15:52:52.232Z [INFO]: # Checking for Git submodules at: /codebuild/output/src932465173/src/via-sem7-bpr2-client/.gitmodules
2023-10-04T15:52:52.240Z [INFO]: # Retrieving environment cache...
2023-10-04T15:52:52.269Z [WARNING]: ! Unable to write cache: {"code":"ERR_BAD_REQUEST","message":"Request failed with status code 404"})}
2023-10-04T15:52:52.270Z [INFO]: ---- Setting Up SSM Secrets ----
2023-10-04T15:52:52.270Z [INFO]: SSM params {"Path":"/amplify/d36cs0wh270jpt/dev/","WithDecryption":true}
2023-10-04T15:53:07.881Z [INFO]: # Patching NPM package '@aws-amplify/cli' from 10.5.2 to 12.5.2...
2023-10-04T15:53:24.242Z [INFO]: # Done patching NPM package '@aws-amplify/cli' to version 12.5.2
2023-10-04T15:53:24.327Z [INFO]: # Retrieving cache...
2023-10-04T15:53:24.366Z [INFO]: # Retrieved cache
2023-10-04T15:53:28.639Z [INFO]: ## Starting Backend Build
                                 # Starting phase: build
2023-10-04T15:53:30.863Z [INFO]: [0mAmplify AppID found: d36cs0wh270jpt. Amplify App name is: viasem7bpr2[0m
2023-10-04T15:53:30.921Z [INFO]: [0mBackend environment dev found in Amplify Console app: viasem7bpr2[0m
2023-10-04T15:53:31.704Z [WARNING]: - Fetching updates to backend environment: dev from the cloud.
2023-10-04T15:53:32.249Z [WARNING]: - Building resource api/bpr2api
2023-10-04T15:53:32.274Z [WARNING]: - Building resource storage/bpr2storage
2023-10-04T15:53:32.285Z [WARNING]: - Building resource storage/bpr2storage2
2023-10-04T15:53:32.293Z [WARNING]: - Building resource storage/bpr2storage3
2023-10-04T15:53:32.367Z [WARNING]: ✔ Successfully pulled backend environment dev from the cloud.
2023-10-04T15:53:32.400Z [WARNING]: Browserslist: caniuse-lite is outdated. Please run:
                                    npx update-browserslist-db@latest
                                    Why you should do it regularly: https://github.com/browserslist/update-db#readme
2023-10-04T15:53:32.419Z [INFO]: ✅
2023-10-04T15:53:34.865Z [INFO]: [33mNote: It is recommended to run this command from the root of your app directory[39m
2023-10-04T15:53:35.654Z [WARNING]: - Initializing your environment: dev
2023-10-04T15:53:36.127Z [WARNING]: - Building resource api/bpr2api
2023-10-04T15:53:36.149Z [WARNING]: - Building resource storage/bpr2storage
2023-10-04T15:53:36.183Z [WARNING]: - Building resource storage/bpr2storage2
2023-10-04T15:53:36.194Z [WARNING]: - Building resource storage/bpr2storage3
2023-10-04T15:53:36.241Z [WARNING]: ✔ Initialized provider successfully.
2023-10-04T15:53:37.178Z [INFO]: [0mNo AppSync API configured. Please add an API[0m
2023-10-04T15:53:42.473Z [WARNING]: ✖ There was an error initializing your environment.
2023-10-04T15:53:42.478Z [INFO]: 🛑 Packaging lambda function failed with the error
                                 Command failed with exit code 127: yarn --no-bin-links --production
                                 error /codebuild/output/src932465173/src/via-sem7-bpr2-client/amplify/backend/function/bpr2lambda1/src/node_modules/bcrypt: Command failed.
                                 Exit code: 127
                                 Command: node-pre-gyp install --fallback-to-build
                                 Arguments:
                                 Directory: /codebuild/output/src932465173/src/via-sem7-bpr2-client/amplify/backend/function/bpr2lambda1/src/node_modules/bcrypt
                                 Output:
                                 /bin/sh: node-pre-gyp: command not found
                                 yarn install v1.22.0
                                 [1/4] Resolving packages...
                                 [2/4] Fetching packages...
                                 [3/4] Linking dependencies...
                                 [4/4] Building fresh packages...
                                 info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
                                 Learn more at: https://docs.amplify.aws/cli/project/troubleshooting/
2023-10-04T15:53:42.485Z [INFO]: 
2023-10-04T15:53:42.485Z [INFO]: Session Identifier: d82c9b99-0890-4030-89d7-e9289073d41d
2023-10-04T15:53:42.486Z [WARNING]: - Creating Zip
2023-10-04T15:53:42.516Z [INFO]: 
2023-10-04T15:53:42.516Z [INFO]: ✅ Report saved: /tmp/viasem7bpr2/report-1696434822493.zip
2023-10-04T15:53:42.517Z [WARNING]: - Sending zip
2023-10-04T15:53:43.832Z [WARNING]: ✔ Done
2023-10-04T15:53:43.833Z [INFO]: Project Identifier: 9f1c3ad01bd453c3cfcda7bec9b89ebe
2023-10-04T15:53:43.858Z [ERROR]: !!! Build failed
2023-10-04T15:53:43.858Z [ERROR]: !!! Non-Zero Exit Code detected
2023-10-04T15:53:43.858Z [INFO]: # Starting environment caching...
2023-10-04T15:53:43.859Z [INFO]: # Environment caching completed
Terminating logging...
```

@rafaelfaria did you manage to find a workaround??

**Comment by NoHop3** [comment]
Edit I made it work with the following workaround -> just install the package that is producing the issue globally prior to any other steps that are written here in this thread...

amplify.yml

```
version: 1
backend:
  phases:
    build:
      commands:
        - '# Execute Amplify CLI with the helper script'
        - yarn global add node-pre-gyp@0.17.0
        - cd amplify/backend/function/bpr2lambda1/src && yarn install && cd -
        - amplifyPush --simple
frontend:
  phases:
    preBuild:
      commands:
        - yarn install --ignore-engines
    build:
      commands:
        - yarn run build
  artifacts:
    baseDirectory: build
    files:
      - '**/*'
  cache:
    paths:
      - node_modules/**/*

```
